etl-pipeline/processes/record_candidate_finder.py

```diff
@@ -0,0 +1,194 @@
+
+        for i in range(0, len(identifiers), batch_size):
+            id_batch = self._format_identifiers(identifiers[i : i + batch_size])
+
+            try:
+                records = (
+                    self.db_manager.session.query(
+                        Record.title, Record.id, Record.identifiers, Record.has_part
+                    )
+                    .filter(~Record.id.in_(already_matched_record_ids))
+                    .filter(Record.identifiers.overlap(id_batch))
+                    .filter(Record.title.isnot(None))
+                    .all()
+                )
+
+                matched_records.extend(records)
+            except DataError:
+                logger.exception("Unable to get matching records")
+
+        return matched_records
+    
+    def _get_records_by_ids(self, record_ids: List[str]) -> List[Record]:
+        return (
+            self.db_manager.session.query(Record)
+            .filter(Record.id.in_(record_ids))
+            .all()
+        )
+    
+    @staticmethod
+    def _titles_overlap(
+        tokenized_record_title: Set[str], tokenized_matched_record_title: Set[str]
+    ) -> bool:
+        """Determines if two titles are similar enough to be considered matching.
+        
+        Rules:
+        1. Single word titles must be subset/superset
+        2. Multi-word titles must share at least 2 words
+        """
+        # Single-word title handling
+        if (
+            len(tokenized_record_title) == 1
+            and not tokenized_record_title <= tokenized_matched_record_title
+        ):
+            return False
+        elif (
+            len(tokenized_matched_record_title) == 1
+            and not tokenized_record_title >= tokenized_matched_record_title
+        ):
+            return False
+        # Multi-word title handling
+        elif (
+            len(tokenized_record_title) > 1 and len(tokenized_matched_record_title) > 1
+        ) and len(tokenized_record_title & tokenized_matched_record_title) < 2:
+            return False
+
+        return True
+    
+    @staticmethod
+    def _tokenize_title(title: str) -> Set[str]:
+        """Converts a title string into a set of normalized "tokens" (words).
+        
+        1. Extracts words using regex
+        2. Converts to lowercase
+        3. Removes common stop words
+        """
+        title_tokens = re.findall(r"(\w+)", title.lower())
+
+        return set(title_tokens) - set(["a", "an", "the", "of"])
+    
+    @staticmethod
+    def _format_identifiers(identifiers: List[str]) -> str:
+        """Formats identifiers for PostgreSQL array overlap query.
+        Returns:
+            Formatted string for Postgres array overlap query
+        """
+        formatted_ids = []
+
+        for id in identifiers:
+            formatted_id = f'"{id}"' if re.search(r"[{},]{1}", id) else id
+            formatted_ids.append(formatted_id)
+
+        return "{{{}}}".format(",".join(formatted_ids)) 
```

Comment:
These methods to me may be long in a separate file! They are also static and don't necessarily need to be part of a class.

Reply:
I figured they should be colocated because I only seem them being used here (and the file size didn't seem too long yet), and then I debated making them free functions but ultimately made them static methods just because they are only really used as part of the class anyway. But a free function in the same file would be fine too. Curious what you prefer?

Reply:
let's iterate on it!